### PR TITLE
Feature/002 resetting of encrypted shared preferences

### PR DESCRIPTION
**Pull request by marvinmirtschin**

### CODE REVIEW

File: rxpreferences/src/main/java/com/nexenio/rxpreferences/provider/EncryptedSharedPreferencesProvider.java [hunk 1]
```diff
@@ -42,4 +67,38 @@ public EncryptedSharedPreferencesProvider(@NonNull SharedPreferences sharedPrefe
...
+     */
+    public Completable deletePreferencesFile(@NonNull Context context) {
+        return Completable.fromAction(() -> {
+            File file = new File(context.getApplicationInfo().dataDir + "/shared_prefs/" + SHARED_PREFERENCES_NAME + ".xml");
+            if (file.exists()) {
+                if (!file.delete()) {
+                    throw new IllegalStateException("Unable to delete preferences file");
+                }
+            }
+            file = new File(context.getApplicationInfo().dataDir + "/shared_prefs/" + SHARED_PREFERENCES_NAME + ".xml.bak");
```
Comment (Steppschuh) [inline rxpreferences/src/main/java/com/nexenio/rxpreferences/provider/EncryptedSharedPreferencesProvider.java]: Duplicated code, can be extracted to new method

File: rxpreferences/src/main/java/com/nexenio/rxpreferences/provider/EncryptedSharedPreferencesProvider.java [hunk 2]
```diff
@@ -42,4 +67,38 @@ public EncryptedSharedPreferencesProvider(@NonNull SharedPreferences sharedPrefe
...
+     * Clear preferences in-memory and in the files. This should be done if the master key used is
+     * lost or if you wish to remove preferences.
+     */
+    public Completable resetPreferences(@NonNull Context context) {
+        return Completable.fromAction(() -> context.getSharedPreferences(SHARED_PREFERENCES_NAME, Context.MODE_PRIVATE)
+                .edit()
+                .clear()
+                .commit())
+                .andThen(deletePreferencesFile(context))
+                .andThen(Completable.fromAction(() -> sharedPreferences = createEncryptedSharedPreferences(context).blockingGet()));
```
Comment (Steppschuh) [inline rxpreferences/src/main/java/com/nexenio/rxpreferences/provider/EncryptedSharedPreferencesProvider.java]: Why the `.blockingGet()`? Can be part of the stream.

File: rxpreferences/src/main/java/com/nexenio/rxpreferences/provider/EncryptedSharedPreferencesProvider.java [hunk 3]
```diff
@@ -42,4 +67,38 @@ public EncryptedSharedPreferencesProvider(@NonNull SharedPreferences sharedPrefe
...
+    }
+
+    /**
+     * Delete persisted shared preferences, so that it can't be restored on the next application
+     * start.
+     */
+    public Completable deletePreferencesFile(@NonNull Context context) {
```
Comment (Steppschuh) [inline rxpreferences/src/main/java/com/nexenio/rxpreferences/provider/EncryptedSharedPreferencesProvider.java]: Is there any use case for deleting the file but keeping the preferences in memory? If not, move the clearing to this method.

File: app/src/androidTest/java/com/nexenio/rxpreferences/provider/TestableEncryptedSharedPreferencesProvider.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+                .andThen(deletePreferencesFile(context))
+                .andThen(Completable.fromAction(() -> sharedPreferences = createEncryptedSharedPreferences(context).blockingGet()));
+    }
+
+    /**
+     * Delete persisted shared preferences, so that it can't be restored on the next application
+     * start.
+     */
+    @Override
+    public Completable deletePreferencesFile(@NonNull Context context) {
```
Comment (Steppschuh) [inline app/src/androidTest/java/com/nexenio/rxpreferences/provider/TestableEncryptedSharedPreferencesProvider.java]: Why the overwrite? Isn't this the same implementation?

File: app/src/androidTest/java/com/nexenio/rxpreferences/provider/EncryptedSharedPreferencesProviderTest.java [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+
+    }
+
+    @Before
+    public void setup() throws KeyStoreException {
+        clear();
+    }
+
+    @After
+    public void cleanUp() throws KeyStoreException {
```
Comment (Steppschuh) [inline app/src/androidTest/java/com/nexenio/rxpreferences/provider/EncryptedSharedPreferencesProviderTest.java]: Why before and after? If you want to clean up after all tests are done, you can use `@AfterClass` instead.

File: app/src/androidTest/java/com/nexenio/rxpreferences/provider/EncryptedSharedPreferencesProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+        clear();
+    }
+
+    // Comment notes: MasterKey available / preferencesFile available / in-memory preferences available
```
Comment (Steppschuh) [inline app/src/androidTest/java/com/nexenio/rxpreferences/provider/EncryptedSharedPreferencesProviderTest.java]: Hard to memorize, would prefer the full text on each test in JavaDoc comment `/**  ... */`

File: app/src/androidTest/java/com/nexenio/rxpreferences/provider/EncryptedSharedPreferencesProviderTest.java [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    @Test
+    public void createEncryptedSharedPreferences_sharedPreferencesFileAndKey_createsPreferences() throws IOException {
+        // create valid key
+        new TestableEncryptedSharedPreferencesProvider(context, SHARED_PREFERENCES_NEW, MASTER_KEY_ALIAS);
+
+        // create valid file
+        File validFile = new File(context.getApplicationInfo().dataDir + "/shared_prefs/" + SHARED_PREFERENCES_NEW + ".xml");
+        File existingFile = new File(context.getApplicationInfo().dataDir + "/shared_prefs/" + SHARED_PREFERENCES_EXISTING + ".xml");
+
+        Scanner myReader = new Scanner(validFile);
```
Comment (Steppschuh) [inline app/src/androidTest/java/com/nexenio/rxpreferences/provider/EncryptedSharedPreferencesProviderTest.java]: Rename to `scanner`

File: app/src/androidTest/java/com/nexenio/rxpreferences/provider/EncryptedSharedPreferencesProviderTest.java [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+
+    private static void removeBlockingInMemoryPreferences() {
+        removeBlockingInMemoryPreference(context.getSharedPreferences(SHARED_PREFERENCES_NEW, Context.MODE_PRIVATE));
+        removeBlockingInMemoryPreference(context.getSharedPreferences(SHARED_PREFERENCES_EXISTING, Context.MODE_PRIVATE));
+    }
+
+    private static void removeBlockingInMemoryPreference(SharedPreferences sharedPreferences) {
+        sharedPreferences
+                .edit()
+                .remove("__androidx_security_crypto_encrypted_prefs_key_keyset__")
```
Comment (Steppschuh) [inline app/src/androidTest/java/com/nexenio/rxpreferences/provider/EncryptedSharedPreferencesProviderTest.java]: Are there constants available for these names? If not you could create them.

File: app/build.gradle [hunk 9]
```diff
@@ -25,9 +25,11 @@ android {
...
-    implementation 'androidx.appcompat:appcompat:1.2.0'
-    testImplementation 'junit:junit:4.13'
-    androidTestImplementation 'androidx.test:runner:1.3.0'
-    androidTestImplementation 'androidx.test.espresso:espresso-core:3.3.0'
+    implementation 'androidx.appcompat:appcompat:1.3.1'
+    implementation 'androidx.security:security-crypto:1.1.0-alpha03'
+    testImplementation 'junit:junit:4.13.1'
+    androidTestImplementation 'androidx.test:runner:1.4.0'
+    androidTestImplementation 'androidx.test.espresso:espresso-core:3.4.0'
+    implementation 'io.reactivex.rxjava3:rxjava:3.1.0'
```
Comment (Steppschuh) [inline app/build.gradle]: Adjust order to not mix `implementation`, `testImplementation` and `androidTestImplementation`

**Review by Steppschuh — CHANGES_REQUESTED** [review]

**Comment by devbeckmann** [comment]
LGTM

**Review by devbeckmann — APPROVED** [review]

**Pull request merged**